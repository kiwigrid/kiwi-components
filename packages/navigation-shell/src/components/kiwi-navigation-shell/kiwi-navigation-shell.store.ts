import { createStore } from '@stencil/store';
import { MaybeAsync } from '../../utils/maybe-async';

// ROUTE

/**
 * The initial route configuration passed to the navigation shell component.
 * It is used to initialize the actual routes.
 */
export type RouteConfig<
  RouteData extends Record<string, unknown>,
  ResolvedRouteData extends RouteData = RouteData
> = {
  routeKey: string;
  label: RouteLabel<RouteData>;
  url: RouteUrl<RouteData>;
  handler: RouteHandler<RouteData>;
  resolver?: RouteDataResolver<RouteData, ResolvedRouteData>;
};

/**
 * The RouteHandler is a function that knows how to perform the actual
 * navigation operation and returns a routes history, which can be seen as the
 * path to a route (normally represented by a breadcrumb in the UI).
 */
interface RouteHandler<D> {
  (data: D): RouteHistory;
}

/**
 * The label describing a route (e.g. for use as link text).
 *
 * Can be either a static string (e.g. `'Show User Profile'`) or a factory function
 * that takes the data associated to a route and returns the label
 * (e.g. ```({ user }) => `Show Profile of ${user}`;```).
 */
type RouteLabel<RouteData> = string | ((data: RouteData) => string);

/**
 * The actual url of this route.
 *
 * Can be either a static string (e.g. `'/users'`) or a factory function that
 * takes the data associated to a route and returns the url (e.g.
 * ```({ userId }) => `/users/${userId}`;```).
 */
type RouteUrl<RouteData> = string | ((data: RouteData) => string);

type RouteDataResolver<RouteData, ResolvedData extends RouteData> = (
  data: RouteData,
) => MaybeAsync<ResolvedData>;

// ROUTE HISTORY

/**
 * The RouteHistory is a list of RouteLinks defining the path to the current
 * Route.
 * It is used to e.g. build the breadcrumb.
 */
export type RouteHistory = RouteLink[];

interface RouteLink {
  label?: string;
  to?: {
    key: string;
    data?: Record<string, unknown> | Promise<Record<string, unknown>>;
  };
}

/**
 * Helper to find the corresponding route to a given key returning its
 * effective url, label and the onClick handler which invokes the navigation
 * handler for this route and updates the breadcrumb (aka RouteHistory).
 *
 * @param key The route key
 * @param data The data associated with this route
 */
export const makeLink = async <RouteData extends Record<string, unknown>>(
  key: string,
  data: RouteData,
): Promise<[url: string, label: string, onClick: (event: Event) => void]> => {
  const route = getRoute(key);

  if (route == undefined) {
    throw new Error(`No handler registered for ${key}`);
  }

  const resolvedData = await Promise.resolve(
    route.resolver?.(data) ?? data,
  ).then((resolved) => ({ ...data, ...resolved }));

  const label =
    typeof route.label === 'function' ? route.label(resolvedData) : route.label;

  const url =
    typeof route.url === 'string' ? route.url : route.url(resolvedData);

  return [
    url,
    label,
    (event: Event) => {
      event.preventDefault();
      state.breadcrumb = [...route.handler(resolvedData), { label }];
      state.activeRoute = key;
    },
  ];
};

// STORE

const { state, dispose, onChange } = createStore<{
  routes: RouteConfig<Record<string, unknown>>[];
  breadcrumb: RouteHistory;
  activeRoute: string;
  routeChangeListeners: ((activeRoute: string) => void)[];
}>({ routes: [], breadcrumb: [], activeRoute: '', routeChangeListeners: [] });

export { state, dispose };

export const init = (
  routes: RouteConfig<Record<string, unknown>>[],
  breadcrumb: RouteHistory,
  activeRoute: string,
): void => {
  state.routes = routes;
  state.breadcrumb = breadcrumb;
  state.activeRoute = activeRoute;
};

onChange('activeRoute', (activeRoute) => {
  Object.values(state.routeChangeListeners).forEach((listener) =>
    listener?.(activeRoute),
  );
});

export const getRoute = (
  key: string,
): RouteConfig<Record<string, unknown>> | undefined =>
  state.routes?.find((route) => route.routeKey === key);

export const hasRoute = (key: string): boolean =>
  state.routes?.some((route) => route.routeKey === key);

export const isActive = (key: string): boolean => state.activeRoute === key;

export const registerRouteChangeListener = (
  listener: (activeRoute: string) => void,
): void => {
  state.routeChangeListeners = [...state.routeChangeListeners, listener];
};

export const unregisterRouteChangeListener = (
  listener: (activeRoute: string) => void,
): void => {
  state.routeChangeListeners = state.routeChangeListeners.filter(
    (l) => l !== listener,
  );
};

export default state;
