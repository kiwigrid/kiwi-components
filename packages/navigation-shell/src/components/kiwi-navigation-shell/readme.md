# kiwi-navigation-shell

## Usage

```jsx
<kiwi-navigation-shell
  routes={{[
    {
      routeKey: 'home',
      label: 'Home',
      url: '/home',
      handler: () => {
        location.href = '/home';

        return [];
      }
    },
    {
      routeKey: 'users',
      label: 'Users',
      url: '/users',
      handler: () => {
        location.href = '/users';

        return [];
      }
    },
    {
      routeKey: 'userProfile',
      label: ({ username }) => `Profile of ${username}`,
      url: ({ username }) => `/users/${username}/profile`,
      handler: ({ username }) => {
        location.href = `/users/${username}/profile`

        // You can define the parents of this route (this route exclusive) here if you want to use the breadcrumb. The current route will be appended automatically.
        return [{ to: { key: 'users' } }]
      }
    },
  ]}}

  // You have to define where you are starting from if you want to use the breadcrumb. You can skip it otherwise.
  breadcrumb={{[

    // assuming we are at `/home` this could look like so (note that if you don't specify the `to` property just the label is rendered, without a link)
    { label: 'Home' } // Will render just 'Home'

    // assuming we are at `/users/johndoe/profile` this could look like so
    { label: 'Users', to: { key: 'users' } }, // Will render link to users
    { label: 'Profile of johndoe' } // Will render label 'Profile of johndoe'
    // resulting in 'Users >> Profile of johndoe'
  ]}}
>
{/** App */}
</kiwi-navigation-shell>
```

Related:

* [Shell Link](../kiwi-shell-link/readme.md)
* [Breadcrumb](../kiwi-shell-breadcrumb/readme.md)

<!-- Auto Generated Below -->


## Properties

| Property              | Attribute      | Description                   | Type                                                              | Default     |
| --------------------- | -------------- | ----------------------------- | ----------------------------------------------------------------- | ----------- |
| `activeRoute`         | `active-route` | Initial active route.         | `string`                                                          | `''`        |
| `breadcrumb`          | --             | Initial breadcrumb.           | `RouteLink[]`                                                     | `[]`        |
| `routes` _(required)_ | --             | Array of route configuration. | `RouteConfig<Record<string, unknown>, Record<string, unknown>>[]` | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
