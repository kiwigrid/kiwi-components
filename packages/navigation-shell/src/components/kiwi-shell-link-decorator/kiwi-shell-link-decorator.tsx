import { Component, Element, h, Host, Prop } from '@stencil/core';
import navShellState, {
  makeLink,
  registerRouteChangeListener,
  unregisterRouteChangeListener,
} from '../kiwi-navigation-shell/kiwi-navigation-shell.store';

@Component({
  tag: 'kiwi-shell-link-decorator',
  shadow: false,
})
export class KiwiShellLinkDecorator {
  /** Route data keyed by route key. Will be passed as route data. */
  @Prop()
  routeData: Record<string, Record<string, unknown>> = {};

  /** Class added to elements marked with `data-shell-active="your-route"` */
  @Prop()
  activeClass = 'active';

  @Element()
  private ref?: HTMLKiwiShellLinkDecoratorElement;

  private unregisterFunctions: (() => void)[] = [];

  componentDidLoad(): void {
    registerRouteChangeListener(this.handleRouteChange);

    this.ref?.querySelectorAll('a[data-shell-route]').forEach((a) => {
      const key = a.getAttribute('data-shell-route');

      if (!key) {
        return;
      }

      const data = this.routeData[key] ?? {};

      try {
        const [url, , onClick] = makeLink(key, data);

        a.setAttribute('href', url);
        a.addEventListener('click', onClick);

        this.unregisterFunctions = [
          ...this.unregisterFunctions,
          () => {
            a.removeEventListener('click', onClick);
          },
        ];
      } catch (e) {
        console.warn(e);
      }
    });

    this.handleRouteChange(navShellState.activeRoute);
  }

  render(): JSX.Element {
    return (
      <Host>
        <slot></slot>
      </Host>
    );
  }

  disconnectedCallback(): void {
    this.unregisterFunctions.forEach((unregister) => unregister());
    unregisterRouteChangeListener(this.handleRouteChange);
  }

  private handleRouteChange = (activeRoute: string): void => {
    this.ref?.querySelectorAll(`[data-shell-active]`).forEach((el) => {
      const route = (el as HTMLElement).dataset?.shellActive;

      if (route === activeRoute) {
        el.classList.add(this.activeClass);
      } else {
        el.classList.remove(this.activeClass);
      }
    });
  };
}
