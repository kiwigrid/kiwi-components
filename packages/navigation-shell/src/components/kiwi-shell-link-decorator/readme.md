# kiwi-shell-link-decorator



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description                                                          | Type                                        | Default    |
| ------------- | -------------- | -------------------------------------------------------------------- | ------------------------------------------- | ---------- |
| `activeClass` | `active-class` | Class added to elements marked with `data-shell-active="your-route"` | `string`                                    | `'active'` |
| `routeData`   | --             | Route data keyed by route key. Will be passed as route data.         | `{ [x: string]: Record<string, unknown>; }` | `{}`       |


## Methods

### `cleanup() => Promise<void>`

Cleanup listeners.

#### Returns

Type: `Promise<void>`




----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
