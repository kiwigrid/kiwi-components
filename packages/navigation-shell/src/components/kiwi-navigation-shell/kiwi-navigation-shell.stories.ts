import { forceReRender } from '@storybook/web-components';
import { html, TemplateResult } from 'lit-html';
import { RouteConfig } from './kiwi-navigation-shell.store';

export default {
  title: 'Navigation Shell',
};

const Root = (): TemplateResult => html`
  <kiwi-shell-link route-key="root.devices"></kiwi-shell-link>
  <kiwi-shell-link route-key="root.customers"></kiwi-shell-link>
  <kiwi-shell-link route-key="root.apps"></kiwi-shell-link>
`;

const Customers = (): TemplateResult => html`
  <p>Customers:
    <ul>
      <li>
        <kiwi-shell-link
          route-key="root.customers.customerDetails"
          .routeData="${{ userId: '1234' }}"
        >
        </kiwi-shell-link>
      </li>
      <li>
        <kiwi-shell-link
          route-key="root.customers.customerDetails"
          .routeData="${{ userId: '5678' }}"
        >
        </kiwi-shell-link>
      </li>
    </ul>
  </p>
`;

const Devices = (): TemplateResult => html`
  <p>Devices:
    <ul>
      <li>
        <kiwi-shell-link
          route-key="root.devices.deviceDetails"
          .routeData="${{ guid: '1234' }}"
        >
        </kiwi-shell-link>
      </li>
      <li>
        <kiwi-shell-link
          route-key="root.devices.deviceDetails"
          .routeData="${{ guid: '5678' }}"
        >
        </kiwi-shell-link>
      </li>
    </ul>
  </p>
`;

const routes = [
  {
    routeKey: 'root',
    label: 'Root',
    url: '/root',
    handler: () => {
      navigateTo(Root);

      return [];
    },
  },
  {
    routeKey: 'root.devices',
    label: 'Devices',
    url: () => `/devices`,
    handler: () => {
      navigateTo(Devices);

      return [{ to: { key: 'root' } }];
    },
  },
  {
    routeKey: 'root.devices.deviceDetails',
    url: ({ guid }) => `/devices/${guid}`,
    label: ({ guid }) => guid,
    handler: ({ guid }) => {
      navigateTo(() => html`<p>Device: ${guid}</p>`);

      return [
        { to: { key: 'root' } },
        { label: 'Devices', to: { key: 'root.devices' } },
      ];
    },
  } as RouteConfig<{ guid: string }>,
  {
    routeKey: 'root.customers',
    label: 'Customers',
    url: () => `/customers`,
    handler: () => {
      navigateTo(Customers);

      return [{ to: { key: 'root' } }];
    },
  },
  {
    routeKey: 'root.customers.customerDetails',
    label: ({ userId }) => userId,
    url: ({ userId }) => `/customers/${userId}`,
    handler: ({ userId }) => {
      navigateTo(() => html`<p>Customer: ${userId}</p>`);

      return [
        { to: { key: 'root' } },
        { label: 'Customers', to: { key: 'root.customers' } },
      ];
    },
  } as RouteConfig<{ userId: string }>,
  {
    routeKey: 'root.apps',
    label: 'Apps',
    url: () => `/apps`,
    handler: () => {
      navigateTo(() => html`<p>Apps</p>`);

      return [{ to: { key: 'root' } }];
    },
  },
];

let content = Root;

const navigateTo = (component: () => TemplateResult): void => {
  content = component;
  forceReRender();
};

const Template = (): TemplateResult => {
  return html`<kiwi-navigation-shell
    .routes="${routes}"
    .breadcrumb="${[{ label: 'Root' }]}"
  >
    <div class="panel panel-default">
      <div class="panel-heading"><h2>Navigation Shell</h2></div>
      <div class="panel-body">
        <kiwi-shell-breadcrumb></kiwi-shell-breadcrumb>
      </div>
    </div>

    <div class="panel panel-default">
      <div class="panel-body">${content()}</div>
    </div>
  </kiwi-navigation-shell>`;
};

export const Basic = Template.bind({});
