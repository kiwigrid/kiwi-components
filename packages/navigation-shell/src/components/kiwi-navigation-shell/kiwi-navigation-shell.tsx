import { Component, ComponentInterface, h, Prop } from '@stencil/core';
import { init, RouteConfig, RouteHistory } from './kiwi-navigation-shell.store';

@Component({
  tag: 'kiwi-navigation-shell',
  shadow: false,
})
export class KiwiNavigationShell implements ComponentInterface {
  /**
   * Array of route configuration.
   *
   * @example
   *
   * [
   *   {
   *     routeKey: 'home',
   *     label: 'Home',
   *     url: '/home',
   *     handler: () => {
   *       location.href = '/home';
   *
   *       return [];
   *     },
   *   }
   * ]
   */
  @Prop()
  public routes!: RouteConfig<
    Record<string, unknown>,
    Record<string, unknown>
  >[];

  /**
   * Initial breadcrumb.
   *
   * @example
   *
   * [
   *   {
   *     label: 'Categories',
   *     to: { key: 'categories' },
   *   },
   *   {
   *     label: 'Household & Kitchen',
   *     to: { key: 'products', data: { categoryId: 'household+kitchen' } },
   *   },
   *   {
   *     label: 'Nicer Dicer',
   *   },
   * ]
   */
  @Prop()
  public breadcrumb: RouteHistory = [];

  /** Initial active route. */
  @Prop()
  public activeRoute = '';

  componentWillLoad(): void {
    init(this.routes, this.breadcrumb, this.activeRoute);
  }

  render(): JSX.Element {
    return <slot></slot>;
  }
}
