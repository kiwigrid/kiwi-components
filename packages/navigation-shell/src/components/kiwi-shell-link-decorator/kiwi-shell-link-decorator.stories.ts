import { html, TemplateResult } from 'lit-html';
import readme from './readme.md';

export default {
  title: 'Shell Link Decorator',
  component: 'kiwi-shell-link-decorator',
  parameters: {
    docs: { description: { component: readme } },
  },
};

const Template: Story<{
  children: TemplateResult;
}> = ({ children }) => {
  return html` <kiwi-navigation-shell
    .routes="${[
      {
        routeKey: 'home',
        label: 'Home',
        url: '/home',
        handler: () => {
          console.log('navigate to home');

          return [];
        },
      },
      {
        routeKey: 'products',
        label: ({ amount }: Record<string, string>) => `${amount} Products`,
        url: '/products',
        resolver: () => Promise.resolve({ amount: '10' }),
        handler: () => {
          console.log('navigate to products');

          return [];
        },
      },
      {
        routeKey: 'contact',
        label: () => 'Contact',
        url: '/contact',
        handler: () => {
          console.log('navigate to contact');

          return [];
        },
      },
    ]}"
    active-route="home"
  >
    <div class="panel panel-default">
      <div class="panel-body">
        <kiwi-shell-link-decorator>${children}</kiwi-shell-link-decorator>
      </div>
    </div>
  </kiwi-navigation-shell>`;
};

export const NavPills: typeof Template = Template.bind({});
NavPills.args = {
  children: html`
    <ul class="nav nav-pills">
      <li data-shell-active="home"><a data-shell-route="home">Home</a></li>
      <li data-shell-active="products">
        <a data-shell-route="products">Products</a>
      </li>
      <li data-shell-active="contact">
        <a data-shell-route="contact">Contact</a>
      </li>
    </ul>
  `,
};
