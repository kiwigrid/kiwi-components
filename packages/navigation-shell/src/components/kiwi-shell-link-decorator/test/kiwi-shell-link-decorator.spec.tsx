import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { init } from '../../kiwi-navigation-shell/kiwi-navigation-shell.store';
import { KiwiShellLinkDecorator } from '../kiwi-shell-link-decorator';

describe('kiwi-shell-link-decorator', () => {
  let handler: jest.Mock;

  beforeAll(async () => {
    init(
      [
        {
          routeKey: 'route-66',
          label: 'Route 66',
          url: () => '/route/66',
          handler: (data) => {
            return handler(data);
          },
        },
        {
          routeKey: 'dynamic-label',
          label: ({ userId }) => `User ${userId}`,
          url: ({ userId }) => `/users/${userId}`,
          handler: (data) => {
            return handler(data);
          },
        },
      ],
      [],
      'route-66',
    );
  });

  beforeEach(() => {
    handler = jest.fn(() => []);
  });

  it('renders', async () => {
    const page = await newSpecPage({
      components: [KiwiShellLinkDecorator],
      template: () => (
        <kiwi-shell-link-decorator
          routeData={{ 'dynamic-label': { userId: 'foo' } }}
        >
          <ul class="nav nav-pills">
            <li data-shell-active="route-66">
              <a data-shell-route="route-66">Home</a>
            </li>
            <li data-shell-active="dynamic-label">
              <a data-shell-route="dynamic-label">Products</a>
            </li>
          </ul>
        </kiwi-shell-link-decorator>
      ),
    });

    expect(page.root).toMatchSnapshot();
  });
});
