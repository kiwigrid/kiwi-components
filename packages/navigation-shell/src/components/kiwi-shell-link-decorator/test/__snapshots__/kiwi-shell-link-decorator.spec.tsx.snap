// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiwi-shell-link-decorator renders 1`] = `
<kiwi-shell-link-decorator>
  <!---->
  <ul class="nav nav-pills">
    <li class="active" data-shell-active="route-66">
      <a data-shell-route="route-66" href="/route/66">
        Home
      </a>
    </li>
    <li data-shell-active="dynamic-label">
      <a data-shell-route="dynamic-label" href="/users/foo">
        Products
      </a>
    </li>
  </ul>
</kiwi-shell-link-decorator>
`;
